// connect-js/queries/filters.js
var Filters;
(function (Filters) {
    var QueryFilter = (function () {
        function QueryFilter(field, operator, value) {
            this.field = field;
            this.operator = operator;
            this.value = value;
        }
        return QueryFilter;
    })();
    Filters.QueryFilter = QueryFilter;
    var QueryFilterBuilder = (function () {
        function QueryFilterBuilder(field) {
            this.field = field;
        }
        QueryFilterBuilder.prototype.eq = function (value) {
            return new Filters.QueryFilter(this.field, 'eq', value);
        };
        QueryFilterBuilder.prototype.ne = function (value) {
            return new Filters.QueryFilter(this.field, 'ne', value);
        };
        QueryFilterBuilder.prototype.gt = function (value) {
            return new Filters.QueryFilter(this.field, 'gt', value);
        };
        QueryFilterBuilder.prototype.gte = function (value) {
            return new Filters.QueryFilter(this.field, 'gte', value);
        };
        QueryFilterBuilder.prototype.lt = function (value) {
            return new Filters.QueryFilter(this.field, 'lt', value);
        };
        QueryFilterBuilder.prototype.lte = function (value) {
            return new Filters.QueryFilter(this.field, 'lte', value);
        };
        QueryFilterBuilder.prototype.exists = function (value) {
            return new Filters.QueryFilter(this.field, 'exists', value);
        };
        QueryFilterBuilder.prototype.startsWith = function (value) {
            return new Filters.QueryFilter(this.field, 'startsWith', value);
        };
        QueryFilterBuilder.prototype.endsWith = function (value) {
            return new Filters.QueryFilter(this.field, 'endsWith', value);
        };
        QueryFilterBuilder.prototype.contains = function (value) {
            return new Filters.QueryFilter(this.field, 'contains', value);
        };
        return QueryFilterBuilder;
    })();
    Filters.QueryFilterBuilder = QueryFilterBuilder;
})(Filters || (Filters = {}));
module.exports = Filters;
