// connect-js/queries/timeframes.js
var Timeframes;
(function (Timeframes) {
    var AbsoluteTimeframe = (function () {
        function AbsoluteTimeframe(start, end) {
            this.start = start;
            this.end = end;
        }
        return AbsoluteTimeframe;
    })();
    Timeframes.AbsoluteTimeframe = AbsoluteTimeframe;
    var RelativeTimeframe = (function () {
        function RelativeTimeframe(isCurrent, length, period) {
            this.isCurrent = isCurrent;
            this.length = length;
            this.period = period;
        }
        RelativeTimeframe.prototype.direction = function () {
            return this.isCurrent ? 'current' : 'previous';
        };
        return RelativeTimeframe;
    })();
    Timeframes.RelativeTimeframe = RelativeTimeframe;
    var TimeframeBuilder = (function () {
        function TimeframeBuilder() {
        }
        TimeframeBuilder.prototype.from = function (from) {
            return new AbsoluteTimeframe(from);
        };
        TimeframeBuilder.prototype.to = function (to) {
            return new AbsoluteTimeframe(null, to);
        };
        TimeframeBuilder.prototype.between = function (start, end) {
            return new AbsoluteTimeframe(start, end);
        };
        TimeframeBuilder.prototype.current = function (length, period) {
            return new RelativeTimeframe(true, length, period);
        };
        TimeframeBuilder.prototype.last = function (length, period) {
            return new RelativeTimeframe(false, length, period);
        };
        return TimeframeBuilder;
    })();
    Timeframes.TimeframeBuilder = TimeframeBuilder;
})(Timeframes || (Timeframes = {}));
module.exports = Timeframes;
